cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/KeyValueTable.java

```diff
@@ -226,7 +225,9 @@ public void write(KeyValue<byte[], byte[]> keyValue) throws IOException {
       private boolean closed = false;
       @Override
       protected KeyValue<byte[], byte[]> computeNext() {
-        Preconditions.checkState(!closed);
+        if (closed) {
```

Comment:
The old logic is not good. Should always use the endOfData to signal end of data instead of throwing illegal state exception.